concordium-consensus/src/Concordium/External/DryRun.hs

```diff
@@ -0,0 +1,881 @@
+        return OK
+
+-- | Mint a specified amount and credit it to the specified account.
+dryRunMintToAccount ::
+    -- | Dry-run handle.
+    StablePtr DryRunHandle ->
+    -- | Account address to mint to.
+    Ptr Word8 ->
+    -- | Amount to mint.
+    Word64 ->
+    -- | Output vector.
+    Ptr ForeignVec ->
+    -- | The return code (see 'DryRunReturnCode').
+    IO Int64
+dryRunMintToAccount dryRunPtr senderPtr mintAmt outVec = dryRunStateHelper dryRunPtr outVec costMintToAccount $
+    \StateHelperInfo{..} -> do
+        sender <- decodeAccountAddress senderPtr
+        runWithEBlockStateContext shiMVR shiBSC $ \drsRef -> do
+            drs@DryRunState{..} <- liftIO $ readIORef drsRef
+            bsoGetAccountIndex drsBlockState sender >>= \case
+                Nothing -> do
+                    liftIO $
+                        writeProtoResponse
+                            shiWriteOut
+                            shiQuotaRem
+                            DryRunErrorAccountNotFound
+                    return OK
+                Just account -> do
+                    bsoMintToAccount drsBlockState account (Amount mintAmt)
+                        >>= liftIO . \case
+                            Left safeMintAmount -> do
+                                writeProtoResponse shiWriteOut shiQuotaRem $
+                                    DryRunErrorAmountOverLimit safeMintAmount
+                                return OK
+                            Right newState -> do
+                                writeIORef drsRef drs{drsBlockState = newState}
+                                writeProtoResponse
+                                    shiWriteOut
+                                    shiQuotaRem
+                                    DryRunSuccessMintedToAccount
+                                return OK
+
+-- | Run a transaction in the current dry-run state, updating the state if it succeeds.
+dryRunTransaction ::
+    -- | Dry run handle.
+    StablePtr DryRunHandle ->
+    -- | Sender account address.
+    Ptr Word8 ->
+    -- | Energy limit for executing the payload
+    Word64 ->
+    -- | Payload.
+    Ptr Word8 ->
+    -- | Payload length.
+    Word64 ->
+    -- | Array of (credential, key) pairs.
+    Ptr Word8 ->
+    -- | Number of signatures.
+    Word64 ->
+    -- | Output vector.
+    Ptr ForeignVec ->
+    -- | The return code (see 'DryRunReturnCode').
+    IO Int64
+dryRunTransaction dryRunPtr senderPtr energyLimit payloadPtr payloadLen sigPairs sigCount outVec =
+    dryRunStateHelper dryRunPtr outVec costTransactionBase $ \StateHelperInfo{..} -> do
+        sender <- decodeAccountAddress senderPtr
+        payloadBytes <- BS.packCStringLen (castPtr payloadPtr, fromIntegral payloadLen)
+        let payload = EncodedPayload (SBS.toShort payloadBytes)
+        sigMap <-
+            if sigCount == 0
+                then return (Map.singleton 0 (Map.singleton 0 Sig.dummySignatureEd25519))
```

Comment:
It's unclear to me where we need this dummy value. For getting the cost we just pass in the number of signatures.

What is preventing us from not having this dummy signature? It's rather ugly.

Reply:
Which signatures are present affects the behaviour of some transactions.